ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,363 @@
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the `/workspace/source` mount point for all containers in the build-pod.
+Shipwright Build Controller also adds it's own helper to ["git clone" the
+repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+In the `BuildRun` resource, the Build Controller will inspect for the following annotations:
+
+- `build.shipwright.io/upload.wait`: boolean attribute, only `true` or `false` accepted;
+- `build.shipwright.io/upload.wait.timeout`: sets the timeout;
+
+The idea is to employ annotations as we would employ flags in a command-line application, they inform
+the Build Controller about the desired changes for that particular `BuildRun` instance.
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  annotations:
+    build.shipwright.io/upload.wait: "true"
+    build.shipwright.io/upload.wait.timeout: "30s"
+spec: {}
+```
+
+Therefore when `build.shipwright.io/wait.upload` is present the waiting will take place, and by
+expressing `build.shipwright.io/upload.wait.timeout` the actual timeout is configured.
+
+Only the `BuildRun` object will accept the "flags" to trigger the waiting for user upload, meaning
+this type of approach is not meant to leak towards the parent `Build` resource.
+
+##### Alternative BuildRun API
+
+The `sources` slice is part of [Remote Artifacts (EP)][RemoteArtifactsEP] and has been worked on, so
+more effort may be needed to create a good API fit. The alternative API is:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+      timeout: 30s
+```
+
+As a alternative to use annotations described before, a new source type `LocalCopy` will be
```

Comment:
I'm rewording this part to explain the tradeoffs of each approach. They could be used in combination, but since the express the same fields, I think we should pick one.

Reply:
ok this to me is an either / or expression ... either all annotations or all api fields 

those are certainly valid options, but I at least think a limited subset of api fields (i.e. just the `type`) followed by annotations is also viable

but between the two you have listed here, if nobody like the mix of annotations and a small subset of api fields like me, then I vote for explicit API fields, and not annotations.  We just then take a conservative approach and minimize the api fields, assuming reasonable defaults for things like timeouts, and get user feedback, which probably leads to use adding things we are thinking about (like timeout) and others we are not.

but to be clear I like 

```yml
---
apiVersion: shipwright.io/v1alpha1
kind: BuildRun
metadata:
  annotations:
    buildrun.shipwright.io/upload.wait: "true"
    buildrun.shipwright.io/upload.wait.timeout: "30s"
spec:
  sources:
    - name: src
      type: LocalCopy
```

where we later on add `    buildrun.shipwright.io/upload.type:` when we have the protoype of something to choose from ... then, when we get confidence in our mutliple choices, we decide how to make that annotation a new field

and if timeout is something that varies widely, we make that an api field ... but if the default we pick works in enough cases, we leave the tuning know as an annotation